File: Go/EX12.go
package main

import (
	"fmt"
)

// Estrutura Produto
type Produto struct {
	nome  string
	preco float64
}

// Construtor para criar um novo produto
func NewProduto(nome string, preco float64) Produto {
	return Produto{nome: nome, preco: preco}
}

// Método para somar dois produtos
func (p Produto) Somar(outro Produto) Produto {
	// Retorna um novo Produto com a soma dos preços
	return Produto{
		nome:  p.nome + " + " + outro.nome,
		preco: p.preco + outro.preco,
	}
}

// Método para exibir informações do produto
func (p Produto) String() string {
	return fmt.Sprintf("Produto: %s, Preço: R$%.2f", p.nome, p.preco)
}

// Função principal
func main() {
	produto1 := NewProduto("Produto A", 80.00)
	produto2 := NewProduto("Produto B", 30.00)

	// Somando os produtos
	produtoTotal := produto1.Somar(produto2)

	// Exibindo o resultado
	fmt.Println(produtoTotal.String()) // Saída: Produto: Produto A + Produto B, Preço: R$110.00
}

File: Java/EX12.java
// Classe Produto
class Produto {
    private String nome;
    private double preco;

    // Construtor
    public Produto(String nome, double preco) {
        this.nome = nome;
        this.preco = preco;
    }

    // Método para somar dois produtos
    public Produto somar(Produto outro) {
        // Retorna um novo Produto com a soma dos preços
        return new Produto(this.nome + " + " + outro.nome, this.preco + outro.preco);
    }

    // Método para exibir informações do produto
    @Override
    public String toString() {
        return "Produto: " + nome + ", Preço: R$" + String.format("%.2f", preco);
    }

    // Método principal para testar a funcionalidade
    public static void main(String[] args) {
        Produto produto1 = new Produto("Produto A", 50.00);
        Produto produto2 = new Produto("Produto B", 30.00);

        // Somando os produtos
        Produto produtoTotal = produto1.somar(produto2);

        // Exibindo o resultado
        System.out.println(produtoTotal);  // Saída: Produto: Produto A + Produto B, Preço: R$80.00
    }
}



